### ISSUE TITLE
Stable Publishing Plugins

### ISSUE THREAD

**Issue by bmuschko** [gradle/gradle#1165]
Users can decide between the old Maven plugin and the "new" Maven publishing plugin. The same goes for ivy vs. ivy-publish. Most of the time this choice is confusing because feature sets and runtime behaviors are different. Remove the "old" Maven plugin. Implement missing features (e.g. signing) in the "new" Maven publishing plugin.

### Expected Behavior

There's one definitive plugin for publishing. Existing bugs are fixed for the plugin. The old plugin is removed from the Gradle code base.

### Current Behavior

Users have to make a choice between two publishing plugin.

**Comment by sebersole** [comment]
As noted multiple places, it is ridiculous that the (perpetually incubating) maven-publish plugin still does not support this.  I'm going to take a stab at plugin-ifying support for this - the Gradle team can do what they want with this (or nothing)... I will have support for it

**Comment by eriwen** [comment]
@sebersole we certainly cannot fault you for being frustrated here. I think the very least we could do is provide some guidance for your plugin and hope it gets us closer to a long-term solution— is that something you or someone from the team can provide @oehme?

**Comment by sebersole** [comment]
I'm actually pretty far along already.  One problem I am hitting is described here : https://discuss.gradle.org/t/access-generatepomfilefor-publication-from-plugin-code/25426

Hopefully I get that figured out, I can push an example of what this is looking like.

**Comment by sebersole** [comment]
To clarify... I am pretty far along doing this as a completely external plugin, using nothing but `gradleApi()`.  This could be made much more concise when/if properly integrated into `MavenPublishingPlugin`

**Comment by jjohannes** [comment]
@sebersole for your information: We will (finally) continue working on this in the next weeks. Targeting Gradle 5.0 at the latest. I can't promise any more concrete delivery date yet, but since this work is also part of a partnership contract we have to fulfil, this will be done this year.

**Comment by sebersole** [comment]
I hope you'll forgive me if I don't hold my breath ;)

Anyway, December 2018 is a loooong time away and yet still "this year" - I need this functionality much sooner.

**Comment by jjohannes** [comment]
@sebersole that's fine of course. I just wanted to let you know. As soon as there is progress from our side, we'll update this issue.

**Comment by sebersole** [comment]
Understood.  Of course if I can't get these details discussed in https://discuss.gradle.org/t/access-generatepomfilefor-publication-from-plugin-code/25426 worked out, I won't be able to go much further with this anyway.

**Comment by sebersole** [comment]
I may end up using Bintray after all, in which case all of this work was not necessary for me :smile:

But in the meantime, here is the working code: https://github.com/sebersole/gradle-clean-room2/tree/master/buildSrc

**Comment by nniesen** [comment]
Thanks @oehme for linking this. I think my [use case in 1061](https://github.com/gradle/gradle/issues/1061#issuecomment-383741897) is similar to the Hibernate projects issue with their 'testing' artifact.

**Comment by marcphilipp** [comment]
@nniesen Can you please create a new issue and include a complete example to reproduce the above problem?

**Comment by marcphilipp** [comment]
**In Gradle 4.8, the Ivy and Maven Publish Plugins will be stable and the preferred option for publishing artifacts.** 🎉

If you're reading this and currently using the old `Upload`-based infrastructure please migrate to the "new" plugins and let us know if there's anything that needs improvement.

**Comment by sebersole** [comment]
Does 4.8 fix any of the short-comings we've been dealing with wrt
publishing?  Things like multiple artifacts for a project...

Or does it simply the "incubating" tag and not address these now
well-known/documented short-comings?

**Comment by hameno** [comment]
Does signing work with the new plugins?

**Comment by oehme** [comment]
@sebersole As mentioned on the other issue, we will tackle some of the shortcomings in 4.9, but for most users the `Upload` infrastructure is no longer necessary.

@hameno Yes, absolutely.

**Comment by hameno** [comment]
@oehme Are you sure?
> Note: Signing the generated POM file generated by this plugin is currently not supported. Future versions of Gradle might add this functionality. Please use the Maven plugin for the purpose of publishing your artifacts to Maven Central.

from: https://docs.gradle.org/current/userguide/publishing_maven.html

**Comment by hameno** [comment]
Okay, I think I'm confused, which plugin exactly should one use for publishing artifacts to a maven repository? maven, maven-publish or the old upload?

**Comment by marcphilipp** [comment]
@hameno Did you notice the version number in the top left corner of the documentation page you've linked (it says 4.7 😉)? 4.8 RC1 will be released this Friday. If you really cannot wait, feel free to download a [nightly build](https://gradle.org/nightly/).

**Comment by hameno** [comment]
@marcphilipp Sorry, did not notice that :) Looking forward to 4.8 then 👍

**Comment by sebersole** [comment]
@oehme Ok.  We'll just skip 4.8 then and wait on 4.9 to see if that bug is fixed there.

**Comment by nniesen** [comment]
Can you put deprecated on the other publishing doc pages and link to the stable one?

**Comment by marcphilipp** [comment]
We will do that eventually (see #4940). For 4.8 we've clearly marked `maven-publish` and `ivy-publish` as the preferred options but we want to fix at least multiple publications bug mentioned above (#1061) before deprecating the old publishing infrastructure.

**Comment by Vampire** [comment]
@hameno `maven` is the "old upload".
If possible I'd always use the new `maven-publish` instead.
Signing the generated publication and especially the generated POM file needed some hacky build script.
You can see how to do it properly up to 4.7 in my next to last comment on https://discuss.gradle.org/t/how-to-publish-artifacts-signatures-asc-files-using-maven-publish-plugin/7422/23?u=vampire and how it is easily done now with a one-liner in my last comment at https://discuss.gradle.org/t/how-to-publish-artifacts-signatures-asc-files-using-maven-publish-plugin/7422/24?u=vampire now that https://github.com/gradle/gradle/issues/4943 got implemented.

**Comment by hameno** [comment]
@marcphilipp Just tried the nightly. When configuring the pom with the new syntax I cannot find a way to set the parent pom. Am I missing something?

**Comment by hameno** [comment]
I'd also like support for mavens settings.xml, especially credentials: https://github.com/gradle/gradle/issues/1236

**Comment by Vampire** [comment]
Use the `withXml` method or some plugin like `me.tatarka.gradle.pom`, which I use because it provides nicer syntax

**Comment by sebersole** [comment]
@hameno As I pointed out on that issue you linked to... I already have a plugin that does this : https://github.com/sebersole/gradle-maven-publish-auth

I completely agree with you that this should be part of Gradle proper

**Comment by oehme** [comment]
Agreed, the POM DSL only supports fields that actually make sense when working with Gradle. Parent POMs don't fit that bill.

**Comment by oehme** [comment]
Let's take any further discussion to the [individual issues](https://github.com/gradle/gradle/labels/in%3Apublishing-plugin) (or create one if some request doesn't exist yet) instead of using this as a bucket list.

**Comment by Vampire** [comment]
@oehme
> Agreed, the POM DSL only supports fields that actually make sense when working with Gradle. Parent POMs don't fit that bill.
But most others do, don't they, especially as couple of them are required for publishing to Sonatype / Maven Central.
I'm talking about things like `url`, `issueManagement`, `developers`, `licenses` and so on.

**Comment by oehme** [comment]
@Vampire Yes, you'll find those on the new DSL.

**Comment by Vampire** [comment]
Which new DSL?
Something that is yet to come? - When?
Or `org.gradle.api.publish.maven.MavenPom`, there I still only see `packaging` and `withXml`.

**Comment by oehme** [comment]
@Vampire I assume you're not looking at the [nightly](https://docs.gradle.org/nightly/javadoc/org/gradle/api/publish/maven/MavenPom.html) docs.

**Comment by Vampire** [comment]
Ah nice, sorry, was looking in my checkout of April 20th.
Didn't expect it to having been changed **that* * recently. :-)
Though I miss `distributionManagement.downloadUrl` @marcphilipp

**Comment by sebersole** [comment]
Easier support for defining relocations would be nice.  We've had to hack
that together a few times now

**Comment by Vampire** [comment]
@sebersole relocation is included, it's the only thing included in `distributionManagement` currently.

**Comment by sebersole** [comment]
Yep, just saw that https://docs.gradle.org/nightly/dsl/org.gradle.api.publish.maven.MavenPomDistributionManagement.html#org.gradle.api.publish.maven.MavenPomDistributionManagement:relocation(org.gradle.api.Action)

**Comment by oehme** [comment]
Yes, we didn't add the `=`-less syntax to the Provider API. From my perspective we shouldn't add it and rather remove it from the old APIs in the long term, since it's one more of those things where you can do things in different ways for no good reason :)

**Comment by Vampire** [comment]
Well, it's for readability sometimes. :-)
